### PR TITLE
Remove sync.Pool from internal/repository

### PR DESCRIPTION
**Pull request by greatroar**
This replaces #2586. Benchmark results:

```
name              old time/op    new time/op    delta
SaveAndEncrypt-8    36.8ms ± 2%    36.9ms ± 2%    ~     (p=0.218 n=10+10)

name              old speed      new speed      delta
SaveAndEncrypt-8   114MB/s ± 2%   114MB/s ± 2%    ~     (p=0.218 n=10+10)

name              old alloc/op   new alloc/op   delta
SaveAndEncrypt-8    21.1MB ± 0%    21.0MB ± 0%  -0.44%  (p=0.000 n=10+10)

name              old allocs/op  new allocs/op  delta
SaveAndEncrypt-8      79.0 ± 0%      77.0 ± 0%  -2.53%  (p=0.000 n=10+10)
```

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by fd0** [comment]
I've just run the Archiver benchmarks for this change:
```
name                     old time/op    new time/op    delta
ArchiverSaveFileSmall-4    2.89ms ± 3%    2.55ms ± 8%  -11.70%  (p=0.008 n=5+5)
ArchiverSaveFileLarge-4     183ms ± 1%     184ms ± 1%     ~     (p=0.421 n=5+5)

name                     old speed      new speed      delta
ArchiverSaveFileSmall-4  1.42MB/s ± 3%  1.61MB/s ± 7%  +13.70%  (p=0.008 n=5+5)
ArchiverSaveFileLarge-4   237MB/s ± 1%   235MB/s ± 1%     ~     (p=0.421 n=5+5)

name                     old alloc/op   new alloc/op   delta
ArchiverSaveFileSmall-4    19.9MB ± 0%    17.9MB ± 0%  -10.13%  (p=0.008 n=5+5)
ArchiverSaveFileLarge-4     199MB ± 3%     238MB ± 0%  +19.57%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
ArchiverSaveFileSmall-4       292 ± 0%       286 ± 0%   -1.85%  (p=0.008 n=5+5)
ArchiverSaveFileLarge-4     1.71k ± 1%     1.71k ± 1%     ~     (p=1.000 n=5+5)
```

So in the benchmarks, the memory usage for larger files is about 20% higher.

For a real-world test I measured max memory usage for archiving:
```
% export RESTIC_REPOSITORY=/tmp/foo RESTIC_PASSWORD=foo

% git describe
v0.9.6-94-gb5c77784

% go build ./cmd/restic

% ./restic init
created restic repository bc4958895f at /tmp/foo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

% xtime ./restic backup ~/shared/work/web
repository bc495889 opened successfully, password is correct
created new cache in /home/fd0/.cache/restic

Files:        3746 new,     0 changed,     0 unmodified
Dirs:            4 new,     0 changed,     0 unmodified
Added to the repo: 1.689 GiB

processed 3746 files, 1.704 GiB in 0:09
snapshot 484feb4a saved
21.21u 3.42s 10.30r 191280kB ./restic backup /home/fd0/shared/work/web

% git checkout no-bufpool
Switched to branch 'no-bufpool'

% git describe
v0.9.6-99-g85f87a2f

% rm -rf /tmp/foo

% go build ./cmd/restic
created restic repository b51eb7db37 at /tmp/foo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

% xtime ./restic backup ~/shared/work/web
repository b51eb7db opened successfully, password is correct
created new cache in /home/fd0/.cache/restic

Files:        3746 new,     0 changed,     0 unmodified
Dirs:            4 new,     0 changed,     0 unmodified
Added to the repo: 1.691 GiB

processed 3746 files, 1.704 GiB in 0:07
snapshot 7d12efbb saved
20.79u 2.85s 7.65r 161256kB ./restic backup /home/fd0/shared/work/web
```

That's ~200MiB with and 161MiB without buffer pool, so in this real-world test memory usage went down!

File: internal/repository/repository.go [hunk 1]
```diff
@@ -225,13 +225,10 @@ func (r *Repository) SaveAndEncrypt(ctx context.Context, t restic.BlobType, data
 
 	debug.Log("save id %v (%v, %d bytes)", id, t, len(data))
 
-	// get buf from the pool
-	ciphertext := getBuf()
-
-	ciphertext = ciphertext[:0]
 	nonce := crypto.NewRandomNonce()
+
+	ciphertext := make([]byte, 0, len(nonce)+len(data)+r.key.Overhead())
```
Comment (fd0) [inline internal/repository/repository.go]: How about this:
```suggestion
	ciphertext := make([]byte, 0, restic.CiphertextLength(len(data)))
```

**Review by fd0 — CHANGES_REQUESTED** [review]
Just a tiny thing, apart from that this PR is good :)

File: internal/repository/repository.go (hunk 1 above)
Comment (greatroar) [reply internal/repository/repository.go]: Force-pushed a new version with this change.

**Comment by greatroar** [comment]
CI checks pass. The change request is still showing up, even though Github knows it's outdated...

**Review by fd0 — APPROVED** [review]
LGTM!

**Pull request merged**